webboot/webboot.go

```diff
@@ -26,32 +33,102 @@ var (
+		"ubuntu":   webboot.Distro{"/boot/vmlinuz ", "/boot/init", "--reuse-cmdline", "https://releases.ubuntu.com/bionic/ubuntu-18.04.2-desktop-amd64.iso"},
 	}
 )
 
+//wgetArgs takes in the arg user passes e.g. name of ISO or URL to an ISO
+//returns an io.ReadCloser, URL link, filename, and err
+//io.ReadCloser and err is returned from wget
+// if the arg is a name of an URL found in the osSystems, the Download Link specificed to the name of ISO is used for wget & urlToName to generate a name
+// and the Download Link then is returned instead of the passed-in URL
+func wgetArgs(arg string) (io.ReadCloser, string, string, error) {
+	var URL, filename string
+	//checks if arg is a key in osSystems map & a valid name to an ISO
+	if u, ok := osSystems[arg]; ok {
+		//if found in map, it finds the ISO's download link to use as URL for wget
+		URL = u.DownloadLink
+		filename = arg
+	} else {
+		//calls urlToName to change the filename to be the base of the link
+		filename = urlToName(arg)
+	}
+	read, err := wget(URL)
+	return read, URL, filename, err
+}
+
+func wget(URL string) (io.ReadCloser, error) {
+	//gets a response from the specified URL
+	resp, err := http.Get(URL)
+	if err != nil {
+		return nil, err
+	}
+	//checks if request to receive URL succeeded and the transmission level is successful on http level
+	if resp.StatusCode != 200 {
+		return nil, fmt.Errorf("non-200 HTTP status: %d", resp.StatusCode)
+	}
+	return resp.Body, nil
+}
+
+//func writer copies from an io.ReadCloser into filename.
+func writer(read io.ReadCloser, filename string) (string, error) {
+	w, err := os.Create(filename)
+	if err != nil {
+		return "", err
+	}
+
+	defer w.Close()
+
+	_, err = io.Copy(w, read)
+	return filename, err
+}
+
+//func urlToName takes URL link & generates a filename using the base of URL link
+// if the URL is empty or URL Path ends in /, urlToName returns index.html
+func urlToName(URL string) string {
+	//if output file is not initialized, argURl is parsed to generate a name for the output file
+	url, err := url.Parse(URL)
+
+	if err != nil {
+		log.Fatalf("Error parsing the URL, got %v, want nil", err)
+	}
+	filename := "index.html"
+	// takes the last part of the url Path as the name of output file. e.g. "archlinux-2019.08.01-x86_64.iso"
+	if url.Path != "" && !strings.HasSuffix(url.Path, "/") {
+		filename = path.Base(url.Path)
+	}
+	return filename
+}
+
+func usage() {
+	log.Printf("Usage: %s [ARGS] URL or name of ISO\n", os.Args[0])
+	flag.PrintDefaults()
+	os.Exit(1)
+}
+
 func main() {
 
 	flag.Parse()
-
+	//allows user to only pass in one argument e.g. "webboot [nameOFISO]" or "webboot [URL]"
+	if flag.NArg() != 1 {
+		usage()
+	}
 	dhclient.Request(*ifName, *timeout, *retry, *verbose, *ipv4, *ipv6)
 
-	if len(flag.Args()) < 1 {
-		log.Fatal("error:pass in an operating system")
+	arg := strings.ToLower(flag.Arg(0))
```

Comment:
what if the user types in "webbot TinyCore" or "webboot tinycore" or "webboot TINYCORE". They should all be treated the same way either all lowercase or uppercase internally to matter how the user writes it.

Reply:
When you would type a URL with MixedCases in the name, it will all get turned into lowercase and treated as lower case letters. If there are lower and upper cases in the URL, yes, I agree as well that you want to be able to treat them different from URL's that do have just lower cases. With the URL converted into lower case letters, it will either give an error to the user saying there is no URL like that printing it out with all lower cases or it will retrieve an actual URL that has the same format that would be all lower cases. In either case, it would not give the user what they exactly asked for. In that case, you do want to be able to use case sensitivity. I will get rid of the lowerCase. That is a good point